Use a better RNG for better shuffle/smart playlists.

kainz opened the pull request:
The default generator guayadeque has used, rand() does not provide
  a good numeric distribution, and the quality of playlist shuffle and
  dynamic playlists suffers as a result.  Add optional use of the
  random() RNG instead.  (This could be made automatic, but I figure it
  should be manual for at least a time, in case someone somehow is
  running guayadeque on a system that can't satisfy POSIX_C_SOURCE, and
  it does change the observed behavior of the RNG.)

kainz committed:
Use a better RNG for better shuffle/smart playlists.

  The default generator guayadeque has used, rand() does not provide
  a good numeric distribution, and the quality of playlist shuffle and
  dynamic playlists suffers as a result.  Add optional use of the
  random() RNG instead.  (This could be made automatic, but I figure it
  should be manual for at least a time, in case someone somehow is
  running guayadeque on a system that can't satisfy POSIX_C_SOURCE, and
  it does change the observed behavior of the RNG.)

thothix commented:
Hi, thanks for your PR.
I'm taking a look on this but I'm not sure there will be any difference. From stdlib:

/* These are the functions that actually do things.  The `random', `srandom',
   `initstate' and `setstate' functions are those from BSD Unices.
   The `rand' and `srand' functions are required by the ANSI standard.
   We provide both interfaces to the same random number generator.  */
/* Return a random long integer between 0 and 2^31-1 inclusive.  */
extern long int random (void) __THROW;

I think we need to use another algorithm/generator like [mt19937](https://cplusplus.com/reference/random/mt19937/) to get a real improvement.

thothix commented:
I got a small test and it's the same indeed.

https://gist.github.com/thothix/077cd5cb35247475c6670e8f3f790464

thothix commented:
The POSIX_C_SOURCE is valid to C Lang, for C++ the random library is available since C++11. As I can see in the test above both random() and srand() executes the same code (has the same results) in C++.
I think it's enough for random playlists and better random number generators are worth to security applications.

Anyway I'm going to implement a mt19937 generator for a matter to keep this knowledge in the code.

kainz commented:
well color me silly! i could have sworn they were different when I tested, but you are 100% correct. Are you thinking of `std::mersenne_twister_engine`, then?  If so, I've updated this PR to do that.

kainz committed:
RNG: implement C++11 mt19937 RNG

Implement use of C++11 mt19337 RNG after discussion and realization that
rand == random in glibc at least.

thothix commented:
I have written a implementation before you updated the PR. I think the best solution is to merge both of our code. I adapted the code to show you some suggestions.

First the cmake: I would use a more generic name to the config (we can change the generator from MT19937 to another one in the future):

option(CXX11_RNG "Enable CXX11 compliant random generator for better RNG" ON)
if (CXX11_RNG)
    add_definitions(-DCXX11_RNG=1)
endif()

thothix commented:
Regarding the RNG it's ok, I would only make a small change to be able to create other generators and be able to improve the seed (with /dev/urandom and custom code).

I did a small refactoring on utils.h.  I think its better to merge you PR and I do the refactoring afterwards.
I'm going to merge it in 0_5_4_beta branch. 
Thanks.

thothix commented:
The PR is to master so if I change it to other branch I think I'll get problems.
I'm not sure about the best way to do it.
I think you can "git merge 0_5_4_beta" into your branch and open a new PR to branch 0_5_4_beta instead of master.

kainz commented:
`git rebase --onto=thothix/0_5_4_beta thothix/master` works fine, and my testing seems to indicate it works as well -- as such I've rebased this PR, and updated the option as you noted above.

kainz committed:
RNG: Change option name to be more generic
